Add motion blur post processing

HydrogenC opened the pull request:
## What does this PR do?
Add motion blur as a post process effect, including a editor component for tuning parameters. Closes #18227.  
A very classical implementation, uses the motion vector texture, which records both camera motions and mesh motions. 
Video showcase: 
https://github.com/user-attachments/assets/6f453ddc-d41f-4a72-bd1e-d7798bcd7eb9



## How was this PR tested?
DX12 on windows

moudgils commented on Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/MotionBlurTileFilter.azsl:
```diff
@@ -0,0 +1,51 @@
+    {
+        return;
+    }
+    
```
Extra line

moudgils commented on Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/MotionBlurTileFilter.azsl:
```diff
@@ -0,0 +1,51 @@
+    for(int i = -1; i <= 1; i++) {
+        [unroll]
+        for(int j = -1; j <= 1; j++) {
+            if (outPixel.x + i >= dimensions.x || outPixel.y + j >= dimensions.y || (int)outPixel.x + i < 0 || (int)outPixel.y + j < 0){
```
bracket down

moudgils commented on Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/MotionBlurTileMaxHor.azsl:
```diff
@@ -0,0 +1,54 @@
+}
+
+[numthreads(8, 8, 1)]
+void MainCS(
```
Add a quick comment about what the shader is doing at a high level. Please go ahead and do the same for other shaders in the PR

moudgils commented on Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/MotionBlurTileMaxHor.azsl:
```diff
@@ -0,0 +1,54 @@
+
+    float2 maxVelocity = float2(0, 0);
+    [unroll]
+    for(int x = 0; x < TILE_SIZE; x++) {
+        if (inPixel.x + x >= inputDimensions.x){
```
bracket down

moudgils commented on Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/MotionBlurTileMaxVer.azsl:
```diff
@@ -0,0 +1,54 @@
+
+    float2 maxVelocity = float2(0, 0);
+    [unroll]
+    for(int y = 0; y < TILE_SIZE; y++) {
+        if (inPixel.y + y >= inputDimensions.y){
```
brackets

moudgils commented on Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/MotionBlur.azsl:
```diff
@@ -0,0 +1,148 @@
+
+    int2 ditherCoord = int2(pixelCoord.x % 2, pixelCoord.y % 2);
+    float samplingDither = (-0.25 + 2.0 * 0.25 * ditherCoord.x) * (-1.0 + 2.0 * ditherCoord.y);
+    // PassSrg::m_outputColor[pixelCoord].rgb = float3(samplingDither, samplingDither, samplingDither);
```
Remove.

HydrogenC commented on Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/MotionBlur.azsl:
```diff
@@ -0,0 +1,148 @@
+    }
+
+    // Calculate noise to make it look better with less samples per pixel
+	float noise = GetGradientNoise(pixelCoord);
```
This noise is unused, dithering is used instead, as suggested in the slides.

HydrogenC commented on Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/MotionBlur.azsl:
```diff
@@ -0,0 +1,148 @@
+    const float pixelToSampleUnitsScale = rsqrt(dot(sampleStep, sampleStep));
+    float4 accumulatedColor = float4(0.0, 0.0, 0.0, 0.0);
+    
+    for (int i = 0; i < iterationCount; ++i)
```
See my comment below

HydrogenC commented:
@moudgils Thanks for spending time reviewing. Now I addressed most of the comments, but in order to unroll the sampling loop, I used shader variants to specify how many samples per pixel in my latest comment. However, I cannot get it to work. Could you offer me some help on what I shall change in [MotionBlurPass.cpp](https://github.com/o3de/o3de/pull/18884/files#diff-b77fc99501545664794fea998e2a362150a3d74322f090e9c3c895e342556481) to make shader varients based on `o_sampleNumber` work? Now it crashes at `GetRenderAttachmentConfiguration()`.

moudgils commented on Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/MotionBlur.azsl:
```diff
@@ -0,0 +1,176 @@
+    int2 ditherCoord = int2(pixelCoord.x % 2, pixelCoord.y % 2);
+    float samplingDither = (-0.25 + 2.0 * 0.25 * ditherCoord.x) * (-1.0 + 2.0 * ditherCoord.y);
+
+    int iterationCount;
+    switch(o_sampleNumber){
+        case SampleNumber::Sample6:
+        {
+            iterationCount = 3;
+            break;
+        }
+        case SampleNumber::Sample14:
+        {
+            iterationCount = 7;
+            break;
+        }
+        case SampleNumber::Sample24:
+        {
+            iterationCount = 12;
+            break;
+        }
+        case SampleNumber::Sample36:
+        {
+            iterationCount = 18;
+            break;
+        }
+    }
```
My apologies. I didnt see that the iterationCount is not hardcoded and passed in. My original comment was meant to simply add the keyword which i see is not possible. My recommendation for now is to remove this new logic around o_sampleNumber and also remove the unroll keyword below. We can add this shader option later with perf numbers to justify the change.

HydrogenC replied:
Got it.

moudgils commented on Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/MotionBlur.azsl:
```diff
@@ -0,0 +1,176 @@
+    float depth = PassSrg::m_linearDepth[pixelCoord].r;
+
+    float2 tileVelocity = PassSrg::m_inputTileVelocity[pixelCoord / uint2(TILE_SIZE, TILE_SIZE)];
+    tileVelocity = ToPixelVelocity(tileVelocity);
```
I think you need to add some noise to tileVelocity. I think the slides mention it too as well as this shader - https://github.com/ValtoGameEngines/CryEngine/blob/d9d2c9f000836f0676e65a90bed40dcc3b1451eb/Engine/Shaders/HWScripts/CryFX/MotionBlur.cfx#L73. Are we able to do the same?

HydrogenC replied:
Oh I missed that, thanks for reminding me.

moudgils reviewed (commented):
This is looking good. Please test Vulkan as well as test with **-rhi-validation=enable** to ensure you are not introducing new driver validation errors. Feel free to ignore any existing ones unrelated to these changes. We want to make sure we are not introducing new validation errors as part of this work. At some point in future I can take a look at testing this work on mac/ios.

nick-l-o3de commented:
> This is looking good. Please test Vulkan as well as test with **-rhi-validation=enable** to ensure you are not introducing new driver validation errors. Feel free to ignore any existing ones unrelated to these changes. We want to make sure we are not introducing new validation errors as part of this work. At some point in future I can take a look at testing this work on mac/ios.

You would have to use `--rhi=vulkan` and `--rhi-validation=enable` to get both

HydrogenC commented:
I tested on Vulkan windows (with validation enabled) and it works the same as DX12. Log contains some errors but no validation errors. 
Log as following: 
<details>
  <summary>Log output</summary>

  <08:07:18> BackupNameAttachment=" Build(0) 23 May 25 (08 07 18)"  -- used by backup system
  <08:07:18> Log Started at Fri May 23 08:07:18 2025
  <08:07:18> Built on Apr 28 2025 23:20:58
  <08:07:18> Running 64 bit Windows version
  <08:07:18> FileVersion: 0.0.0.0
  <08:07:18> ProductVersion: 0.0.0.0
  <08:07:18> Using Microsoft (tm) C++ Standard Library implementation
  <08:07:18> Project Path: C:\Users\x1398\O3DE\Projects\PrTestProj
  
  <08:07:18> Project Name: PrTestProj
  
  <08:07:18> > cl_assetLoadWarningEnable : true
  
  <08:07:18> > cl_assetLoadWarningMsThreshold : 100
  
  <08:07:18> Loading Config file game.cfg (C:/Users/x1398/O3DE/Projects/PrTestProj/Cache/pc/game.cfg)
  <08:07:18> creating CVarGroups from directory 'Config/CVarGroups' ...
  <08:07:18> Loading Config file user.cfg (C:/Users/x1398/O3DE/Projects/PrTestProj/Cache/pc/config/user.cfg)
  <08:07:18> GameName: PrTestProj
  <08:07:18> BuildTime: Apr 28 2025 23:20:58
  <08:07:18> > g_languageAudio : en-US
  
  <08:07:18> Running without any Audio Engine!
  
  <08:07:18> Audio System is initialized and ready!
  
  <08:07:18> Initializing additional systems
  
  <08:07:18> Using FreeType 2.11.1
  <08:07:18> > LoadLevel : defaultlevel
  
  <08:07:18> Current Language: Chinese (Simplified) 
  <08:07:18> Microsoft Windows 11 家庭中文版|C:\WINDOWS|\Device\Harddisk0\Partition3 10.0.26100
  <08:07:18> Local time is 08:07:18 05/23/25, system running for 4 minutes
  <08:07:18> 32387MB phys. memory installed, 11018MB paging available
  <08:07:18> Current display mode is 2560x1600x32, (Unknown graphics card)
  <08:07:18> IBM enhanced (101/102-key) keyboard and 5+ button mouse installed
  <08:07:18> --------------------------------------------------------------------------------
  <08:07:19> (RPI system) - System assets initialized
  
  <08:07:19> project_path = C:\Users\x1398\O3DE\Projects\PrTestProj
  <08:07:19> sys_localization_folder = Localization
  <08:07:19> Console Variable r_MeasureOverdraw not declared
  <08:07:19> [Plugin Manager] Loading plugins...
  <08:07:19> [Plugin Manager] Successfully loaded plugin 'ComponentEntityEditor', version '1' (GUID: {11B0041C-BC34-4827-A3E4-AB7458FFF678})
  <08:07:19> FFMPEG plugin: CreatePluginInstance
  <08:07:19> FFMPEG plugin: Failed to execute FFmpeg. Please install FFmpeg.
  <08:07:19> [Plugin Manager] Successfully loaded plugin 'FFMPEG Writer', version '1' (GUID: {D2A3A44A-00FF-4341-90BA-89A473F44A65})
  <08:07:19> [Plugin Manager] Successfully loaded plugin 'ProjectSettingsTool', version '1' (GUID: {C5B96A1A-036A-46F9-B7F0-5DF93494F988})
  <08:07:20> [Plugin Manager] Successfully loaded plugin 'QtAssetImporter', version '1' (GUID: {0abf28f2-ef56-4ac9-a459-175abb40d649})
  <08:07:20> (EMotionFX) - -----------------------------------------------
  
  <08:07:20> (EMotionFX) - EMotion Studio Core - Information
  
  <08:07:20> (EMotionFX) - -----------------------------------------------
  
  <08:07:20> (EMotionFX) - Compilation date: Apr 29 2025
  
  <08:07:20> (EMotionFX) - -----------------------------------------------
  
  <08:07:20> (EMotionFX) - EMotion Studio initialized...
  
  <08:07:28> > r_enablePerMeshShaderOptionFlags : false
  
  <08:07:28> > r_meshInstancingEnabled : false
  
  <08:07:28> > r_meshInstancingEnabledForTransparentObjects : false
  
  <08:07:28> > r_meshInstancingBucketSortScatterBatchSize : 512
  
  <08:07:28> [Warning] (Az Console) - attempted to assign out of range value -16
  <08:07:50> [Warning] XML reader: Can't open file (C:/Users/x1398/O3DE/Projects/PrTestProj/user/Sandbox/Macros.xml)
  <08:07:50> Document created
  <08:07:50> Adding '*$tmp*' to ignore masks for changed files.
  <08:07:52> [Warning] (Application) - 'IsPrefabSystemEnabled' is deprecated, the editor only supports prefabs for level editing.
  <08:07:52> Engine initialized, took 44.57s.
  <08:07:52> [CONSOLE] Executing console command 'exec editor_autoexec.cfg'
  <08:07:52> Executing console batch file (try game,config,root): "@products@/editor_autoexec.cfg" not found!
  <08:07:52> [CONSOLE] Executing console command 'exec editor.cfg'
  <08:07:52> Executing console batch file (try game,config,root): "@products@/editor.cfg" not found!
  <08:07:52> [CONSOLE] Executing console command 'exec user.cfg'
  <08:07:52> Executing console batch file (try game,config,root): "user.cfg" found in game/config/ ...
  <08:07:53> (RHI) - ****************************************************************
  
  <08:07:53> (RHI) -                     Registering vulkan RHI                          
  
  <08:07:53> (RHI) - ****************************************************************
  
  <08:07:53> > r_enablePerMeshShaderOptionFlags : false
  
  <08:07:53> > r_meshInstancingEnabled : false
  
  <08:07:53> > r_meshInstancingEnabledForTransparentObjects : false
  
  <08:07:53> > r_meshInstancingBucketSortScatterBatchSize : 512
  
  <08:07:53> [Warning] (Az Console) - attempted to assign out of range value 0
  <08:07:55> Opening level levels/default/default.spawnable
  <08:07:55> Loading from C:/Users/x1398/O3DE/Projects/PrTestProj/Levels/Default/Default.prefab...
  <08:07:55> ---- Start: Game Engine level load
  <08:07:55> Loading map 'C:/Users/x1398/O3DE/Projects/PrTestProj/Levels/Default' into engine...
  <08:07:55> ---- End: Game Engine level load (0 seconds)
  <08:07:55> ---- Start: Post Load
  <08:07:55> ---- End: Post Load (0 seconds)
  <08:07:55> [LevelLoadTime] Level C:/Users/x1398/O3DE/Projects/PrTestProj/Levels/Default loaded in 0 seconds
  <08:07:55> -----------------------------------------------------------
  <08:07:55> Successfully opened document C:/Users/x1398/O3DE/Projects/PrTestProj/Levels/Default/Default.prefab
  <08:07:55> Level loading time: 0.08 seconds
  <08:07:55> -----------------------------------------------------------
  <08:07:55> [Warning] (FrameGraph) - Invalid State: attachment 'Root.MainPipeline_0.PipelineOutput' was added but never used!
  <08:08:22> [Warning] (Model) - ray intersection against a model that is still creating spatial information
  <08:09:21> [Warning] (RPI) - Failed to initialize PipelineState for DynamicDrawContext
  <08:09:21> [Warning] (RPI) - Failed to initialize PipelineState for DynamicDrawContext
  <08:09:21> [Warning] (RPI) - Failed to initialize PipelineState for DynamicDrawContext
  <08:09:21> [Warning] (RPI) - Failed to initialize PipelineState for DynamicDrawContext

</details>

moudgils commented:
> I tested on Vulkan windows (with validation enabled) and it works the same as DX12. Log contains some warnings but no validation errors. Log as following:
> 
> Log output

As long as the warnings are present without motion blur I approve.

nick-l-o3de commented:
```
In file included from /data/workspace/o3de/Gems/AtomLyIntegration/CommonFeatures/Code/Include/AtomLyIntegration/CommonFeatures/PostProcess/MotionBlur/MotionBlurComponentConfig.h:11:
 In file included from /data/workspace/o3de/Gems/Atom/Feature/Common/Code/Include/Atom/Feature/PostProcess/MotionBlur/MotionBlurSettingsInterface.h:11:
 /data/workspace/o3de/Gems/Atom/Feature/Common/Code/Include/Atom/Feature/PostProcess/MotionBlur/MotionBlurConstants.h:18:30: error: unknown type name 'u32'
             static constexpr u32 DefaultSampleNumber = 5;
                             ^
```

consider using AZ::u32 anywhere you have previously used u32.

HydrogenC commented:
> ```
> In file included from /data/workspace/o3de/Gems/AtomLyIntegration/CommonFeatures/Code/Include/AtomLyIntegration/CommonFeatures/PostProcess/MotionBlur/MotionBlurComponentConfig.h:11:
>  In file included from /data/workspace/o3de/Gems/Atom/Feature/Common/Code/Include/Atom/Feature/PostProcess/MotionBlur/MotionBlurSettingsInterface.h:11:
>  /data/workspace/o3de/Gems/Atom/Feature/Common/Code/Include/Atom/Feature/PostProcess/MotionBlur/MotionBlurConstants.h:18:30: error: unknown type name 'u32'
>              static constexpr u32 DefaultSampleNumber = 5;
>                              ^
> ```
> 
> consider using AZ::u32 anywhere you have previously used u32 in c++ files. We use AZ::xxx (u32, u16, etc).

Done. I changed all the occurrences, but some code are copied from other existing source files and they don't use `AZ::u32` over `u32`.

HydrogenC commented:
> gah its now complaining
> 
> ```
> [2025-06-18T22:56:47.129Z] /data/workspace/o3de/Gems/Atom/Feature/Common/Code/Include/Atom/Feature/PostProcess/MotionBlur/MotionBlurConstants.h:18:34: error: no type named 'u32' in namespace 'AZ'
> [2025-06-18T22:56:47.129Z]             static constexpr AZ::u32 DefaultSampleNumber = 5;
> ```
> 
> What is the correct AZ declaration for an unsigned 32 bit integer? Let me take a look at the source code.
> 
> It is AZ::u32, but you have to include `<AzCore\base.h>`

Oh, I'm sorry I didn't notice that. Now I've ensured that it compiles successfully.
